gflownet.py

```diff
@@ -576,49 +647,41 @@ def sample_batch(
+                        env,
+                        policy="model",
+                        model=model,
+                        temperature=self.temperature_logits,
                     )
                     n_actions += 1
             envs = envs[n_empirical:]
-        # Rest of batch
+        # Policy trajectories
         while envs:
-            states = [env.state2obs() for env in envs]
-            mask = [env.no_eos_mask() for env in envs]
-            random_action = self.rng.uniform()
-            if train is False or random_action > self.random_action_prob:
-                with torch.no_grad():
-                    t0_a_model = time.time()
-                    action_probs = model(tf(states))
-                    if self.mask_eos:
-                        action_probs[mask, -1] = -1000
-                    t1_a_model = time.time()
-                    times["actions_model"] += t1_a_model - t0_a_model
-                    if all(torch.isfinite(action_probs).flatten()):
-                        actions = Categorical(logits=action_probs).sample()
-                    else:
-                        random_action = -1
-                        if self.debug:
-                            print("Action could not be sampled from model!")
-            # TODO: review: random actions sampled for all envs!
-            # TODO: review: if random from numpy, tl(x) gives weird results for scalars
-            if train and random_action < self.random_action_prob:
-                high = (len(self.env.action_space) + 1) * np.ones(len(envs), dtype=int)
-                if self.mask_eos:
-                    high[mask] -= 1
-                actions = self.rng.integers(low=0, high=high, size=len(envs))
+            # Forward sampling
+            if train is False:
+                envs, actions, valids = self.forward_sample(
+                    envs, times, policy="model", model=model, temperature=1.0
+                )
+            else:
+                envs, actions, valids = self.forward_sample(
+                    envs,
+                    times,
+                    policy="model",
+                    model=model,
+                    temperature=self.temperature_logits,
+                )
             t0_a_envs = time.time()
-            assert len(envs) == actions.shape[0]
-            for env, action in zip(envs, actions):
-                state, action, valid = env.step(action)
+            # Add to batch
+            for env, action, valid in zip(envs, actions, valids):
                 if valid:
-                    parents, parents_a = env.get_parents(state, env.done)
+                    parents, parents_a = env.get_parents()
                     if train:
                         batch.append(
                             [
                                 tf([env.state2obs()]),
-                                tl(action),
-                                state,
+                                tl([action]),
```

Comment:
tl([action]) is not used later, is it ?

Reply:
I believe it is used at least by `unpack_terminal_states()`